packages/devtools_app/lib/src/blocking_action_mixin.dart

```diff
@@ -19,17 +21,37 @@ mixin BlockingActionMixin<T extends StatefulWidget> on State<T> {
   bool get actionInProgress => _actionInProgress;
   bool _actionInProgress = false;
 
-  void blockWhileInProgress(Future callback()) async {
+  final _disposed = Completer<bool>();
+
+  /// Sets actionInProgress to true until the [callback] completes or this
+  /// State object is disposed.
+  ///
+  /// The future returned by this method completes when either the future
+  /// returned by the callback completes or the State object is disposed.
+  Future<void> blockWhileInProgress(Future callback()) async {
     setState(() {
       _actionInProgress = true;
     });
     try {
-      // TODO(jacobr): handle actions that timeout gracefully.
-      await callback();
+      // If we await a callback that does not complete until after this object
+      // is disposed, we will leak State objects.
+      await Future.any([callback(), _disposed.future]);
```

Comment:
should we block the `callback()` from being called if `_disposed.future` completes first? `Future.any` doesn't stop the other futures from completing.

Reply:
There isn't a way to cancel the in progress `callback()` if _disposed completes first .

Reply:
You can't cancel the callback but you can prevent it from running with a boolean flag.